improc/simulator.py

```diff
@@ -0,0 +1,886 @@
+
+        return y
+
+    def get_star_flux_values(self):
+        """
+        Return the fluxes of the stars (in photons per total exposure time)
+        after possibly applying a flux change due to e.g., occultations/flares,
+        or due to scintillation noise.
+        (TODO: this is not yet implemented!)
+
+        """
+        return self.star_mean_fluxes
+
+
+class Simulator:
+    """
+    Make simulated images for testing image processing techniques.
+    """
+
+    def __init__(self, **kwargs):
+        self.pars = SimPars(**kwargs)
+
+        # classes holding parts of the simulation
+        self.sensor = None
+        self.camera = None
+        self.sky = None
+        self.stars = None
+
+        # holds the truth values for this image
+        self.truth = None
+
+        # intermediate variables
+        # fluxes coming from the stars
+        self.star_x = None
+        self.star_y = None
+        self.star_f = None
+
+        self.psf = None  # we are cheating because this includes both the optical and atmospheric PSFs
+        self.oversampling = None  # how much oversampling (integer) do we need to express fluxes?
+        self.flux_top = None  # this is the mean number of photons hitting the top of the atmosphere
+
+        # adding the sky into the mix
+        self.flux_with_sky = None  # this is mean photons after passing through the atmosphere, adding the background
+
+        # adding the effect of the camera optics
+        self.flux_vignette = None  # average number of photons after passing through the aperture vignette
+
+        # now the photons are absorbed into the sensor pixels
+        self.electrons = None  # average number of electrons in each pixel, considering QE
+
+        self.average_counts = None  # the final counts, not including noise
+        self.noise_var_map = None  # the total variance from read, dark, sky b/g, and source noise
+
+        # outputs:
+        self.image = None
+
+    def make_sensor(self):
+        """
+        Generate a sensor and save it to the simulator.
+        This includes all the properties of the pixels
+        and amplifiers and readout electronics.
+
+        Usually this does not change between images
+        from the same survey.
+        """
+        self.sensor = SimSensor()
+
+        self.sensor.bias_mean = self.pars.bias_mean
+        self.sensor.pixel_bias_std = self.pars.bias_std
+        # self.sensor.pixel_bias_map = self.sensor.bias_mean + np.random.poisson(
+        #     self.sensor.pixel_bias_std ** 2, size=self.pars.imsize
+        # ) - self.sensor.pixel_bias_std ** 2
+        self.sensor.pixel_bias_map = np.random.normal(
+            self.sensor.bias_mean,
+            self.sensor.pixel_bias_std,
+            size=self.pars.imsize
+        )
+
+        self.sensor.gain_mean = self.pars.gain_mean
+        self.sensor.pixel_gain_std = self.pars.gain_std
+        self.sensor.pixel_gain_map = np.random.normal(
```

Comment:
Not suggesting a change, but musing : are the pixel gains on a typical real CCD distributed normally? I suspect not exactly, and I also suspect they're not all uncorrelated.